### PR TITLE
feat(changelog-tool): add changelog command line tool

### PR DESCRIPTION
**Pull request by vm-001**

### CODE REVIEW

File: changelog.lua [hunk 1]
```diff
@@ -0,0 +1,374 @@
+local pl_template = require "pl.template"
+local pl_tablex = require "pl.tablex"
+local pl_file = require "pl.file"
+local pl_dir = require "pl.dir"
+local pl_path = require "pl.path"
+local pl_stringx = require "pl.stringx"
+local lyaml = require "lyaml"
+local pl_app = require 'pl.lapp'
+local cjson = require "cjson"
+local http = require("socket.http")
```
Comment (dndx) [inline changelog.lua]: Please do not use `lua-http`, this library had significant quality issues and we had to remove it from our codebase completely.

File: .github/workflows/generate.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    steps:
+      - uses: actions/checkout@v4
+
+      - uses:  actions/checkout@v4
+        with:
+          repository: Kong/kong
+          ref: test/changelog-remove-prs-and-jiras
+          path: kong
+
+      - uses: leafo/gh-actions-lua@v10
```
Comment (dndx) [inline .github/workflows/generate.yml]: I suggest we use an official OpenResty release to run this. The CI should be using same technology developers use on their own machines or we risk not discovering it no longer works on our own machine until too late.

File: schema.json [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    },
+    "issues": {
+      "type": "array",
+      "description": "List of associated GitHub issues",
+      "items": {
+        "pattern": "^(\\d+|\\w+\/\\w+:\\d+)$",
+        "type": ["integer", "string"],
+        "examples": ["1", "torvalds/linux:1"]
+      }
+    },
```
Comment (dndx) [inline schema.json]: Is this really necessary? It seems that:

https://github.com/Kong/kong/pull/11568
https://github.com/Kong/kong/issues/11568

both goes to the same page, we should not need to distinguish them to such detail.

File: changelog.lua [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+  end,
+}
+
+
+local args = pl_app [[
+Usage: changelog <command> [options]
+
+Commands:
+  release <version>                  release a release note based on the files in the CHANGELOG/unreleased directory.
+  preview <version>                  preview a release note based on the files in the CHANGELOG/unreleased directory.
```
Comment (dndx) [inline changelog.lua]: Please explain what's the difference of `preview` and `release`.

File: changelog.lua [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+      if is_empty_table(entry.prs) then
+        entry.prs = { ctx.pull.number }
+      end
+      if is_empty_table(entry.jiras) then
+        local jiras = {}
+
+        local sources = { string.upper(ctx.commit.message), string.upper(ctx.pull.body) }
+        for _, text in ipairs(sources) do
+          match_pattern(text, "FTI%-%d+", jiras)
+          match_pattern(text, "KAG%-%d+", jiras)
```
Comment (dndx) [inline changelog.lua]: We have other projects. Suggest we do not hard code these prefixes.

File: changelog.lua [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+
+  return sorted_files
+end
+
+
+local function fetch_commit_context(filename)
+  local ctx = {}
+  filename = string.sub(filename, 6)
+  --print(filename)
+  local commit_url = "https://api.github.com/repos/kong/kong/commits?path=%s"
```
Comment (dndx) [inline changelog.lua]: Should the `kong` repo name be configurable? We have other repos we might use as well.

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,6 @@
+install_dependencies:
+	luarocks install lua-cjson
+	luarocks install penlight
+	luarocks install lyaml
+	luarocks install luasocket
+	luarocks install luasec
```
Comment (dndx) [inline Makefile]: Again, `luasocket` and `luasec` are unmaintained and unreliable libraries. Please do not use them.

File: changelog.lua [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+
+local function parse_github_ref(system, reference_type, references)
+  if references == nil or references == lyaml.null then
+    return nil
+  end
+  local parsed_references = {}
+  for i, ref in ipairs(references or {}) do
+    local repo = REPOS[system]
+    local ref_no = tonumber(ref) -- treat ref as number string first
+    local name = "#" .. ref
```
Comment (dndx) [inline changelog.lua]:
```suggestion
    local name = "#" .. ref

```

File: changelog.lua [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+    if not ref_no then -- ref is not a number string
+      local parts = pl_stringx.split(ref, ":")
+      repo = parts[1]
+      ref_no = parts[2]
+      name = pl_stringx.replace(tostring(ref), ":", " #")
+    end
```
Comment (dndx) [inline changelog.lua]:
```suggestion
    end

```

Try to keep a consistent style to how `Kong/kong`'s code is written. It makes reading code easier too.

File: changelog.lua [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+      id = jira,
+      link = JIRA_BASE_URL .. jira
+    }
+  end
+  return jira_items
+end
+
+
+local function is_yaml(filename)
+  return pl_stringx.endswith(filename, ".yaml") or
```
Comment (dndx) [inline changelog.lua]: Remove the `.yaml` support.

File: changelog.lua [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+    pl_stringx.endswith(filename, ".yml")
+end
+
+local function is_empty_table(t)
+  return t == nil or next(t) == nil
```
Comment (dndx) [inline changelog.lua]: Use https://github.com/openresty/luajit2/tree/v2.1-agentzh#tableisempty after PUC Lua is get rid of.

File: schema.json [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    "issues": {
+      "type": "array",
+      "description": "List of associated GitHub issues",
+      "items": {
+        "pattern": "^(\\d+|\\w+\/\\w+:\\d+)$",
+        "type": ["integer", "string"],
+        "examples": ["1", "torvalds/linux:1"]
```
Comment (dndx) [inline schema.json]: I don't think we will ever support linking PRs outside of the org. AFAIK no one has used the `torvalds/linux:1` format up until now. Is this really needed?

**Review by dndx — CHANGES_REQUESTED** [review]

File: changelog.lua [hunk 13]
```diff
@@ -0,0 +1,374 @@
...
+  data.version = version
+  local content = assert(generate_content(data))
+  local target_path = pl_path.join(path, version)
+
+  if pl_path.exists(target_path) and version ~= folder then
+    error("directory exists, please manually remove. " .. version)
+  end
+
+  if folder == UNRELEASED then
+    os.execute("mv " .. pl_path.join(path, UNRELEASED) .. " " .. target_path)
```
Comment (dndx) [inline changelog.lua]: These does not really need to be so automatic, I would not feel too comfortable having a script `mv` file on the disk for me, it is very easy to make a mistake in the command and cause damage. 

Just print out the command and let user execute it themself.

I even argue this script should not even write the `.md` file directly, use the KISS principal and let user do:

`./changelog.lua 3.4.0. /path/to/kong/unreleased > CHANGELOG-3.4.0.md`

is much better and safer.

If we really need to automate, Makefile and a shell script is much better idea.

**Review by dndx — CHANGES_REQUESTED** [review]

File: .github/workflows/generate.yml (hunk 2 above)
Comment (vm-001) [reply .github/workflows/generate.yml]: Removed. This file is only for test purpose

File: schema.json (hunk 3 above)
Comment (vm-001) [reply schema.json]: Done in https://github.com/Kong/gateway-changelog/pull/2

File: schema.json (hunk 12 above)
Comment (vm-001) [reply schema.json]: Done in https://github.com/Kong/gateway-changelog/pull/2

File: changelog.lua (hunk 5 above)
Comment (vm-001) [reply changelog.lua]: https://github.com/Kong/kong/pull/11613. If the rule is too broad, then how can we avoid matching Sep-2023 ?

File: changelog.lua (hunk 10 above)
Comment (vm-001) [reply changelog.lua]: I will keep both file suffixes support here until one of them is completely gone.

File: Makefile [hunk 14]
```diff
@@ -0,0 +1,9 @@
+test:
```
Comment (vm-001) [inline Makefile]: Only for demo, will be removed.

**Review by dndx — APPROVED** [review]

**Pull request merged**